src/sql_generator.cpp

```diff
@@ -0,0 +1,312 @@
+                 std::vector<column_def> columns) {
+
+    std::set<std::string> alter_types;
+    std::set<std::string> added_columns;
+    std::set<std::string> deleted_columns;
+
+    auto existing_columns = describe_table(con, db_name, schema_name, table_name);
+    std::map<std::string, column_def> new_column_map;
+
+    for (auto &col: columns) {
+        new_column_map.emplace(col.name, col);
+        added_columns.emplace(col.name);
+    }
+
+    for (auto &col: existing_columns) {
+        auto new_col_it = new_column_map.find(col.name);
+
+        added_columns.erase(col.name);
+
+        if (new_col_it == new_column_map.end()) {
+            deleted_columns.emplace(col.name);
+        } else if (new_col_it->second.type != col.type) {   // altering primary key not supported in duckdb
+            alter_types.emplace(col.name);
+        }
+    }
+
+    for (const auto &col_name: added_columns) {
+        std::ostringstream out;
+        out << "ALTER TABLE " + tablename(db_name, schema_name, table_name) << " ADD COLUMN ";
+        auto col = new_column_map.find(col_name)->second;
+
+        out << KeywordHelper::WriteQuoted(col_name, '"') << " " << duckdb::EnumUtil::ToChars(col.type);
+        if (col.primary_key) {
+            out << " PRIMARY KEY";
+        }
+        auto query = out.str();
+        mdlog::info("alter_table: " + query);
+
+        auto result = con.Query(query);
+        if (result->HasError()) {
+            throw std::runtime_error(result->GetError());
+        }
+    }
+
+    for (const auto &col_name: deleted_columns) {
+        std::ostringstream out;
+        out << "ALTER TABLE " << tablename(db_name, schema_name, table_name) << " DROP COLUMN ";
+        auto col = new_column_map.find(col_name)->second;
+
+        out << KeywordHelper::WriteQuoted(col_name, '"');
+
+        auto query = out.str();
+        mdlog::info("alter_table: " + query);
+        auto result = con.Query(query);
+        if (result->HasError()) {
+            throw std::runtime_error(result->GetError());
+        }
+    }
+
+    for (const auto &col_name: alter_types) {
+        std::ostringstream out;
+        out << "ALTER TABLE " << tablename(db_name, schema_name, table_name) << " ALTER ";
+        auto col = new_column_map.find(col_name)->second;
+
+        out << KeywordHelper::WriteQuoted(col_name, '"') << " TYPE " << duckdb::EnumUtil::ToChars(col.type);
+
+        auto query = out.str();
+        auto result = con.Query(query);
+        if (result->HasError()) {
+            throw std::runtime_error(result->GetError());
+        }
+    }
+}
+
+
+void upsert(duckdb::Connection &con, std::string db_name, std::string schema_name, std::string table_name,
+            std::string staging_table_name, const std::vector<std::string> primary_keys,
+            std::vector<column_def> columns) {
+    std::ostringstream sql;
+    sql << "INSERT INTO " + tablename(db_name, schema_name, table_name)
```

Comment:
```suggestion
    sql << "INSERT INTO " << tablename(db_name, schema_name, table_name)
```

Reply:
yeah, those are left over from old concatenation before I switched to the stream. Thanks!